File: cpp/color_refine.cpp
// Color refinement (1-WL) benchmark
// Input: graph file (n m, then m edges u v)
// Output: read=Xms compute=Yms rounds=R colors=C checksum=K
#include <cstdio>
#include <cstdlib>
#include <cstdint>
#include <cstring>
#include <chrono>
#include <algorithm>
#include <unordered_map>
#include <vector>

using Clock = std::chrono::steady_clock;

static double ms_since(Clock::time_point t0) {
    return std::chrono::duration<double, std::milli>(Clock::now() - t0).count();
}

int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "Usage: %s <graph_file>\n", argv[0]); return 1; }

    // ── Read ─────────────────────────────────────────────────────────────────
    auto t0 = Clock::now();

    FILE *f = fopen(argv[1], "r");
    if (!f) { perror("fopen"); return 1; }

    uint32_t n, m;
    fscanf(f, "%u %u", &n, &m);

    // CSR-style adjacency: offsets + flat neighbor array
    // First pass: count degrees
    std::vector<uint32_t> deg(n, 0);
    std::vector<uint32_t> eu(m);
    std::vector<uint32_t> ev(m);
    for (uint32_t i = 0; i < m; i++) {
        fscanf(f, "%u %u", &eu.at(i), &ev.at(i));
        deg.at(eu.at(i))++;
        deg.at(ev.at(i))++;
    }
    fclose(f);

    // Build adjacency arrays
    std::vector<uint32_t> offset(n + 1);
    offset.at(0) = 0;
    for (uint32_t i = 0; i < n; i++) offset.at(i + 1) = offset.at(i) + deg.at(i);
    uint32_t total_edges = offset.at(n);
    std::vector<uint32_t> adj(total_edges);
    std::vector<uint32_t> pos(n, 0);
    for (uint32_t i = 0; i < m; i++) {
        uint32_t u = eu.at(i), v = ev.at(i);
        adj.at(offset.at(u) + pos.at(u)++) = v;
        adj.at(offset.at(v) + pos.at(v)++) = u;
    }

    double read_ms = ms_since(t0);

    // ── Compute ──────────────────────────────────────────────────────────────
    auto t1 = Clock::now();

    std::vector<uint32_t> color(n, 0);
    std::vector<uint32_t> new_color(n);
    std::vector<uint64_t> sig_hash(n);
    // Reusable buffer for neighbor colors
    uint32_t max_deg = 0;
    for (uint32_t v = 0; v < n; v++) {
        uint32_t d = offset.at(v + 1) - offset.at(v);
        if (d > max_deg) max_deg = d;
    }
    std::vector<uint32_t> nbuf(max_deg);
    std::unordered_map<uint64_t, uint32_t> mapping;
    mapping.reserve(n);

    uint32_t rounds = 0;
    uint32_t num_colors = 0;

    for (uint32_t round = 0; round < n; round++) {
        // Build signature hash for each vertex
        for (uint32_t v = 0; v < n; v++) {
            uint32_t lo = offset.at(v), hi = offset.at(v + 1);
            uint32_t deg_v = hi - lo;
            for (uint32_t i = 0; i < deg_v; i++)
                nbuf.at(i) = color.at(adj.at(lo + i));
            std::sort(nbuf.begin(), nbuf.begin() + deg_v);

            uint64_t h = (uint64_t)color.at(v) * 1000003ULL;
            h = h ^ ((uint64_t)deg_v * 2654435761ULL);
            h = h * 1000003ULL;
            for (uint32_t i = 0; i < deg_v; i++) {
                h = h ^ ((uint64_t)nbuf.at(i) * 2654435761ULL);
                h = h * 1000003ULL;
            }
            sig_hash.at(v) = h;
        }

        // Map hashes to consecutive colors, track stability
        mapping.clear();
        uint32_t next_id = 0;
        bool stable = true;
        for (uint32_t v = 0; v < n; v++) {
            auto it = mapping.find(sig_hash.at(v));
            uint32_t cid;
            if (it == mapping.end()) {
                cid = next_id;
                mapping[sig_hash.at(v)] = next_id++;
            } else {
                cid = it->second;
            }
            new_color.at(v) = cid;
            if (cid != color.at(v)) stable = false;
        }

        rounds = round + 1;
        num_colors = next_id;

        if (stable) break;

        std::swap(color, new_color);
    }

    // Compute checksum
    int64_t checksum = 0;
    for (uint32_t v = 0; v < n; v++) checksum += color.at(v);

    double compute_ms = ms_since(t1);

    // ── Output ───────────────────────────────────────────────────────────────
    printf("read=%.1fms compute=%.1fms rounds=%u colors=%u checksum=%lld\n",
           read_ms, compute_ms, rounds, num_colors, (long long)checksum);

    return 0;
}

File: rust/src/bin/color_refine.rs
// Color refinement (1-WL) benchmark
use std::collections::HashMap;
use std::env;
use std::fs;
use std::time::Instant;

#[inline(always)]
fn parse_usize(bytes: &[u8]) -> usize {
    let mut n: usize = 0;
    for &b in bytes {
        n = n * 10 + (b - b'0') as usize;
    }
    n
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <graph_file>", args[0]);
        std::process::exit(1);
    }

    // ── Read ─────────────────────────────────────────────────────────────────
    let t0 = Instant::now();

    let bytes = fs::read(&args[1]).unwrap();
    let mut iter = bytes
        .split(|&b| b == b' ' || b == b'\n' || b == b'\r' || b == b'\t')
        .filter(|s| !s.is_empty());
    let n: usize = parse_usize(iter.next().unwrap());
    let m: usize = parse_usize(iter.next().unwrap());

    // Build CSR adjacency
    let mut deg = vec![0u32; n];
    let mut eu = Vec::with_capacity(m);
    let mut ev = Vec::with_capacity(m);
    for _ in 0..m {
        let u: usize = parse_usize(iter.next().unwrap());
        let v: usize = parse_usize(iter.next().unwrap());
        deg[u] += 1;
        deg[v] += 1;
        eu.push(u);
        ev.push(v);
    }

    let mut offset = vec![0u32; n + 1];
    for i in 0..n {
        offset[i + 1] = offset[i] + deg[i];
    }
    let total = offset[n] as usize;
    let mut adj = vec![0u32; total];
    let mut pos = vec![0u32; n];
    for i in 0..m {
        let u = eu[i];
        let v = ev[i];
        let pu = offset[u] + pos[u];
        adj[pu as usize] = v as u32;
        pos[u] += 1;
        let pv = offset[v] + pos[v];
        adj[pv as usize] = u as u32;
        pos[v] += 1;
    }

    let read_ms = t0.elapsed().as_secs_f64() * 1000.0;

    // ── Compute ──────────────────────────────────────────────────────────────
    let t1 = Instant::now();

    let mut color = vec![0u32; n];
    let mut new_color = vec![0u32; n];
    let mut sig_hash = vec![0u64; n];

    // Find max degree for reusable buffer
    let mut max_deg: usize = 0;
    for v in 0..n {
        let d = (offset[v + 1] - offset[v]) as usize;
        if d > max_deg {
            max_deg = d;
        }
    }
    let mut nbuf = vec![0u32; max_deg];
    let mut mapping: HashMap<u64, u32> = HashMap::with_capacity(n);

    let mut rounds: u32 = 0;
    let mut num_colors: u32 = 0;

    for round in 0..n {
        // Build signature hash for each vertex
        for v in 0..n {
            let lo = offset[v] as usize;
            let hi = offset[v + 1] as usize;
            let deg_v = hi - lo;
            for i in 0..deg_v {
                let w = adj[lo + i] as usize;
                nbuf[i] = color[w];
            }
            nbuf[..deg_v].sort_unstable();

            let mut h: u64 = (color[v] as u64).wrapping_mul(1000003);
            h ^= (deg_v as u64).wrapping_mul(2654435761);
            h = h.wrapping_mul(1000003);
            for &nc in &nbuf[..deg_v] {
                h ^= (nc as u64).wrapping_mul(2654435761);
                h = h.wrapping_mul(1000003);
            }
            sig_hash[v] = h;
        }

        // Map hashes to consecutive colors, track stability
        mapping.clear();
        let mut next_id: u32 = 0;
        let mut stable = true;
        for v in 0..n {
            let h = sig_hash[v];
            let id = mapping.entry(h).or_insert_with(|| {
                let id = next_id;
                next_id += 1;
                id
            });
            new_color[v] = *id;
            if *id != color[v] {
                stable = false;
            }
        }

        rounds = (round + 1) as u32;
        num_colors = next_id;

        if stable {
            break;
        }

        std::mem::swap(&mut color, &mut new_color);
    }

    // Compute checksum
    let checksum: i64 = color.iter().map(|&c| c as i64).sum();

    let compute_ms = t1.elapsed().as_secs_f64() * 1000.0;

    println!(
        "read={:.1}ms compute={:.1}ms rounds={} colors={} checksum={}",
        read_ms, compute_ms, rounds, num_colors, checksum
    );
}
